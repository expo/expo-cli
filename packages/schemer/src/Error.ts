import ExtendableError from 'es6-error';

export class SchemerError extends ExtendableError {
  errors: Array<ValidationError>;
  constructor(errors: Array<ValidationError>) {
    const message = errors.map(e => e.message).join('\n');
    super(message);
    this.errors = errors;
  }
}

export class ValidationError extends ExtendableError {
  errorCode: string;
  fieldPath: string;
  message: string;
  data: any;
  meta: any;
  constructor({
    errorCode,
    fieldPath,
    message,
    data,
    meta,
  }: {
    errorCode: ErrorCode;
    fieldPath: string;
    message: string;
    data: any;
    meta: any;
  }) {
    super(message);
    this.errorCode = errorCode;
    this.fieldPath = fieldPath;
    this.message = message;
    this.data = data;
    this.meta = meta;
  }
}

export type ErrorCode = keyof typeof ErrorCodes;

export const ErrorCodes = {
  SCHEMA_VALIDATION_ERROR: 'SCHEMA_VALIDATION_ERROR',
  SCHEMA_ADDITIONAL_PROPERTY: 'SCHEMA_ADDITIONAL_PROPERTY',
  SCHEMA_MISSING_REQUIRED_PROPERTY: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
  SCHEMA_INVALID_PATTERN: 'SCHEMA_INVALID_PATTERN',

  INVALID_ASSET_URI: 'INVALID_ASSET_URI',
  INVALID_DIMENSIONS: 'INVALID_DIMENSIONS',
  INVALID_CONTENT_TYPE: 'INVALID_CONTENT_TYPE',
  NOT_SQUARE: 'NOT_SQUARE',
};
