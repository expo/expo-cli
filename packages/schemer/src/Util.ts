import fill from 'lodash/fill';
import flatten from 'lodash/flatten';
import get from 'lodash/get';
import zip from 'lodash/zip';

export const fieldPathToSchemaPath = (fieldPath: string) => {
  let newPath = zip(fill(fieldPath.split('.'), 'properties'), fieldPath.split('.'));
  return flatten(newPath).join('.');
};
// Assumption: used only for jsonPointer returned from traverse
export const schemaPointerToFieldPath = (jsonPointer: string) => {
  return jsonPointer
    .split('/')
    .slice(2)
    .filter(e => e !== 'properties')
    .join('.');
};

export const fieldPathToSchema = (schema: Object, fieldPath: string) => {
  return get(schema, fieldPathToSchemaPath(fieldPath));
};
