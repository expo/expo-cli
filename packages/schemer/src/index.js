/**
 * @flow
 **/

import 'babel-polyfill';
import _ from 'lodash';
import Ajv from 'ajv';
import path from 'path';
import fs from 'fs';
import traverse from 'json-schema-traverse';
import readChunk from 'read-chunk';
import imageProbe from 'probe-image-size';
import { SchemerError, ValidationError, ErrorCodes } from './Error';
import { schemaPointerToFieldPath, fieldPathToSchema } from './Util';

type Options = {
  allErrors?: boolean,
  rootDir?: string,
  verbose?: boolean,
  format?: 'full' | 'empty', //figure out later
  metaValidation?: boolean,
};

type Meta = {
  asset?: boolean,
  dimensions?: Object,
  square: boolean,
  contentTypePattern: string,
  contentTypeHuman: string,
};

export { SchemerError, ValidationError, ErrorCodes } from './Error';
export default class Schemer {
  options: Options;
  ajv: Object;
  schema: Object;
  rootDir: string;
  manualValidationErrors: Array<ValidationError>;
  // Schema is a JSON Schema object
  constructor(schema: Object, options: Options = {}) {
    this.options = _.extend(
      {
        allErrors: true,
        verbose: true,
        format: 'full',
        metaValidation: true,
      },
      options
    );

    this.ajv = new Ajv(this.options);
    this.schema = schema;
    this.rootDir = this.options.rootDir || __dirname;
    this.manualValidationErrors = [];
  }

  _formatAjvErrorMessage({
    keyword,
    dataPath,
    params,
    parentSchema,
    data,
    message,
  }: {
    keyword: string,
    dataPath: string,
    params: Object,
    parentSchema: Object,
    data: any,
    message: string,
  }) {
    // This removes the "." in front of a fieldPath
    dataPath = dataPath.slice(1);
    switch (keyword) {
      case 'additionalProperties':
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_ADDITIONAL_PROPERTY,
          fieldPath: dataPath,
          message: `should NOT have additional property '${params.additionalProperty}'`,
          data,
          meta: parentSchema.meta,
        });
      case 'required':
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_MISSING_REQUIRED_PROPERTY,
          fieldPath: dataPath,
          message: `is missing required property '${params.missingProperty}'`,
          data,
          meta: parentSchema.meta,
        });
      case 'pattern': {
        //@TODO Parse the message in a less hacky way. Perhaps for regex validation errors, embed the error message under the meta tag?
        const regexHuman = _.get(parentSchema, 'meta.regexHuman');
        const regexErrorMessage = regexHuman
          ? `'${dataPath}' should be a ${regexHuman[0].toLowerCase() + regexHuman.slice(1)}`
          : `'${dataPath}' ${message}`;
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_INVALID_PATTERN,
          fieldPath: dataPath,
          message: regexErrorMessage,
          data,
          meta: parentSchema.meta,
        });
      }
      default:
        return new ValidationError({
          errorCode: ErrorCodes.SCHEMA_VALIDATION_ERROR,
          fieldPath: dataPath,
          message,
          data,
          meta: parentSchema.meta,
        });
    }
  }

  getErrors(): Array<ValidationError> {
    // Convert AJV JSONSchema errors to our ValidationErrors
    let valErrors = [];
    if (this.ajv.errors) {
      valErrors = this.ajv.errors.map(e => this._formatAjvErrorMessage(e));
    }
    const bothErrors = _.concat(valErrors, this.manualValidationErrors);
    return bothErrors;
  }

  _throwOnErrors() {
    // Clean error state after each validation
    const errors = this.getErrors();
    if (errors.length > 0) {
      this.manualValidationErrors = [];
      this.ajv.errors = [];
      throw new SchemerError(errors);
    }
  }

  async validateAll(data: any) {
    await this._validateSchemaAsync(data);
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateAssetsAsync(data: any) {
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateSchemaAsync(data: any) {
    await this._validateSchemaAsync(data);
    this._throwOnErrors();
  }

  _validateSchemaAsync(data: any) {
    this.ajv.validate(this.schema, data);
  }

  async _validateAssetsAsync(data: any) {
    let assets = [];
    traverse(this.schema, { allKeys: true }, (subSchema, jsonPointer, a, b, c, d, property) => {
      if (property && subSchema.meta && subSchema.meta.asset) {
        const fieldPath = schemaPointerToFieldPath(jsonPointer);
        assets.push({
          fieldPath,
          data: _.get(data, fieldPath),
          meta: subSchema.meta,
        });
      }
    });
    await Promise.all(assets.map(this._validateAssetAsync.bind(this)));
  }

  async _validateAssetAsync({
    fieldPath,
    data,
    meta,
  }: {
    fieldPath: string,
    data: string,
    meta: Meta,
  }) {
    if (meta && meta.asset && data) {
      const { asset, dimensions, square, contentTypePattern, contentTypeHuman }: Meta = meta;
      // filePath could be an URL
      const filePath = path.resolve(this.rootDir, data);
      try {
        // Assumption: All assets are images. This may change in the future.
        // NOTE(nikki): The '4100' below should be enough for most file types, though we
        //              could probably go shorter....
        //              http://www.garykessler.net/library/file_sigs.html
        //  The metadata content for .jpgs might be located a lot farther down the file, so this
        //  may pose problems in the future.
        //  This cases on whether filePath is a remote URL or located on the machine
        const probeResult = fs.existsSync(filePath)
          ? imageProbe.sync(await readChunk(filePath, 0, 4100))
          : await imageProbe(data, { useElectronNet: false });

        const { width, height, type, mime, wUnits, hUnits } = probeResult;

        if (contentTypePattern && !mime.match(new RegExp(contentTypePattern))) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.INVALID_CONTENT_TYPE,
              fieldPath,
              message: `field '${fieldPath}' should point to ${
                meta.contentTypeHuman
              } but the file at '${data}' has type ${type}`,
              data,
              meta,
            })
          );
        }

        if (dimensions && (dimensions.height !== height || dimensions.width != width)) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.INVALID_DIMENSIONS,
              fieldPath,
              message: `'${fieldPath}' should have dimensions ${dimensions.width}x${
                dimensions.height
              }, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }

        if (square && width !== height) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: ErrorCodes.NOT_SQUARE,
              fieldPath,
              message: `image should be square, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }
      } catch (e) {
        this.manualValidationErrors.push(
          new ValidationError({
            errorCode: ErrorCodes.INVALID_ASSET_URI,
            fieldPath,
            message: `cannot access file at '${data}'`,
            data,
            meta,
          })
        );
      }
    }
  }

  async validateProperty(fieldPath: string, data: any) {
    const subSchema = fieldPathToSchema(this.schema, fieldPath);
    this.ajv.validate(subSchema, data);

    if (subSchema.meta && subSchema.meta.asset) {
      await this._validateAssetAsync({ fieldPath, data, meta: subSchema.meta });
    }
    this._throwOnErrors();
  }

  validateName(name: string) {
    return this.validateProperty('name', name);
  }

  validateSlug(slug: string) {
    return this.validateProperty('slug', slug);
  }

  validateSdkVersion(version: string) {
    return this.validateProperty('sdkVersion', version);
  }

  validateIcon(iconPath: string) {
    return this.validateProperty('icon', iconPath);
  }
}
