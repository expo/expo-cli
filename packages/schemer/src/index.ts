import path from 'path';
import fs from 'fs';

import _ from 'lodash';
import Ajv from 'ajv';
import imageProbe from 'probe-image-size';
import readChunk from 'read-chunk';
import traverse from 'json-schema-traverse';

import { schemaPointerToFieldPath, fieldPathToSchema } from './Util';
import { SchemerError, ValidationError } from './Error';

type Options = {
  allErrors?: boolean;
  rootDir?: string;
  verbose?: boolean;
  format?: 'full' | 'empty'; //figure out later
  metaValidation?: boolean;
};

type Meta = {
  asset?: boolean;
  dimensions?: {
    width: number;
    height: number;
  };
  square?: boolean;
  contentTypePattern?: string;
  contentTypeHuman?: string;
};

type AssetField = { fieldPath: string; data: string; meta: Meta };

export { SchemerError, ValidationError, ErrorCodes, ErrorCode } from './Error';
export default class Schemer {
  options: Options;
  ajv: Ajv.Ajv;
  schema: Object;
  rootDir: string;
  manualValidationErrors: Array<ValidationError>;
  // Schema is a JSON Schema object
  constructor(schema: Object, options: Options = {}) {
    this.options = _.extend(
      {
        allErrors: true,
        verbose: true,
        format: 'full',
        metaValidation: true,
      },
      options
    );

    this.ajv = new Ajv(this.options);
    this.schema = schema;
    this.rootDir = this.options.rootDir || __dirname;
    this.manualValidationErrors = [];
  }

  _formatAjvErrorMessage({
    keyword,
    dataPath,
    params,
    parentSchema,
    data,
    message,
  }: Ajv.ErrorObject) {
    const meta = parentSchema && (parentSchema as any).meta;
    // This removes the "." in front of a fieldPath
    dataPath = dataPath.slice(1);
    switch (keyword) {
      case 'additionalProperties': {
        return new ValidationError({
          errorCode: 'SCHEMA_ADDITIONAL_PROPERTY',
          fieldPath: dataPath,
          message: `should NOT have additional property '${(params as any).additionalProperty}'`,
          data,
          meta,
        });
      }
      case 'required':
        return new ValidationError({
          errorCode: 'SCHEMA_MISSING_REQUIRED_PROPERTY',
          fieldPath: dataPath,
          message: `is missing required property '${(params as any).missingProperty}'`,
          data,
          meta,
        });
      case 'pattern': {
        //@TODO Parse the message in a less hacky way. Perhaps for regex validation errors, embed the error message under the meta tag?
        const regexHuman = _.get(parentSchema, 'meta.regexHuman');
        const regexErrorMessage = regexHuman
          ? `'${dataPath}' should be a ${regexHuman[0].toLowerCase() + regexHuman.slice(1)}`
          : `'${dataPath}' ${message}`;
        return new ValidationError({
          errorCode: 'SCHEMA_INVALID_PATTERN',
          fieldPath: dataPath,
          message: regexErrorMessage,
          data,
          meta,
        });
      }
      default:
        return new ValidationError({
          errorCode: 'SCHEMA_VALIDATION_ERROR',
          fieldPath: dataPath,
          message: message || 'Validation error',
          data,
          meta,
        });
    }
  }

  getErrors(): ValidationError[] {
    // Convert AJV JSONSchema errors to our ValidationErrors
    let valErrors: ValidationError[] = [];
    if (this.ajv.errors) {
      valErrors = this.ajv.errors.map(e => this._formatAjvErrorMessage(e));
    }
    const bothErrors = _.concat(valErrors, this.manualValidationErrors);
    return bothErrors;
  }

  _throwOnErrors() {
    // Clean error state after each validation
    const errors = this.getErrors();
    if (errors.length > 0) {
      this.manualValidationErrors = [];
      this.ajv.errors = [];
      throw new SchemerError(errors);
    }
  }

  async validateAll(data: any) {
    await this._validateSchemaAsync(data);
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateAssetsAsync(data: any) {
    await this._validateAssetsAsync(data);
    this._throwOnErrors();
  }

  async validateSchemaAsync(data: any) {
    await this._validateSchemaAsync(data);
    this._throwOnErrors();
  }

  _validateSchemaAsync(data: any) {
    this.ajv.validate(this.schema, data);
  }

  async _validateAssetsAsync(data: any) {
    let assets: AssetField[] = [];
    traverse(this.schema, { allKeys: true }, (subSchema, jsonPointer, a, b, c, d, property) => {
      if (property && subSchema.meta && subSchema.meta.asset) {
        const fieldPath = schemaPointerToFieldPath(jsonPointer);
        assets.push({
          fieldPath,
          data: _.get(data, fieldPath),
          meta: subSchema.meta,
        });
      }
    });
    await Promise.all(assets.map(this._validateAssetAsync.bind(this)));
  }

  async _validateImageAsync({ fieldPath, data, meta }: AssetField) {
    if (meta && meta.asset && data) {
      const { dimensions, square, contentTypePattern }: Meta = meta;
      // filePath could be an URL
      const filePath = path.resolve(this.rootDir, data);
      try {
        // NOTE(nikki): The '4100' below should be enough for most file types, though we
        //              could probably go shorter....
        //              http://www.garykessler.net/library/file_sigs.html
        //  The metadata content for .jpgs might be located a lot farther down the file, so this
        //  may pose problems in the future.
        //  This cases on whether filePath is a remote URL or located on the machine
        const probeResult = fs.existsSync(filePath)
          ? imageProbe.sync(await readChunk(filePath, 0, 4100))
          : await imageProbe(data, { useElectronNet: false });
        if (!probeResult) {
          return;
        }

        const { width, height, type, mime } = probeResult;

        if (contentTypePattern && !mime.match(new RegExp(contentTypePattern))) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'INVALID_CONTENT_TYPE',
              fieldPath,
              message: `field '${fieldPath}' should point to ${
                meta.contentTypeHuman
              } but the file at '${data}' has type ${type}`,
              data,
              meta,
            })
          );
        }

        if (dimensions && (dimensions.height !== height || dimensions.width !== width)) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'INVALID_DIMENSIONS',
              fieldPath,
              message: `'${fieldPath}' should have dimensions ${dimensions.width}x${
                dimensions.height
              }, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }

        if (square && width !== height) {
          this.manualValidationErrors.push(
            new ValidationError({
              errorCode: 'NOT_SQUARE',
              fieldPath,
              message: `image should be square, but the file at '${data}' has dimensions ${width}x${height}`,
              data,
              meta,
            })
          );
        }
      } catch (e) {
        this.manualValidationErrors.push(
          new ValidationError({
            errorCode: 'INVALID_ASSET_URI',
            fieldPath,
            message: `cannot access file at '${data}'`,
            data,
            meta,
          })
        );
      }
    }
  }

  async _validateAssetAsync({ fieldPath, data, meta }: AssetField) {
    if (meta && meta.asset && data) {
      if (meta.contentTypePattern && meta.contentTypePattern.startsWith('^image')) {
        await this._validateImageAsync({ fieldPath, data, meta });
      }
    }
  }

  async validateProperty(fieldPath: string, data: any) {
    const subSchema = fieldPathToSchema(this.schema, fieldPath);
    this.ajv.validate(subSchema, data);

    if (subSchema.meta && subSchema.meta.asset) {
      await this._validateAssetAsync({ fieldPath, data, meta: subSchema.meta });
    }
    this._throwOnErrors();
  }

  validateName(name: string) {
    return this.validateProperty('name', name);
  }

  validateSlug(slug: string) {
    return this.validateProperty('slug', slug);
  }

  validateSdkVersion(version: string) {
    return this.validateProperty('sdkVersion', version);
  }

  validateIcon(iconPath: string) {
    return this.validateProperty('icon', iconPath);
  }
}
