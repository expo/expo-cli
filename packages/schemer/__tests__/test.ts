import { ErrorCodes, SchemerError } from '../src/Error';
import Schemer from '../src/index';
import good from './files/app.json';
import bad from './files/bad.json';
import badWithNot from './files/badwithnot.json';
import invalidAppIcon from './files/invalidAppIcon.json';
import schema from './files/schema.json';

const S = new Schemer(schema.schema, { rootDir: './__tests__' });

describe('Sanity Tests', () => {
  it('is a class', () => {
    const schema = require('./files/schema.json');
    const S = new Schemer(schema, { rootDir: './__tests__' });
    expect(S instanceof Schemer).toBe(true);
  });

  it('has public functions', () => {
    const schema = require('./files/schema.json');
    const S = new Schemer(schema, { rootDir: './__tests__' });
    expect(S.validateAll).toBeDefined();
    expect(S.validateProperty).toBeDefined();
  });
});

describe('Image Validation', () => {
  it('errors for webp images', async () => {
    let didError = false;
    try {
      await S.validateAssetsAsync({
        android: {
          adaptiveIcon: { foregroundImage: './files/webp.webp' },
        },
      });
    } catch (e) {
      didError = true;
      expect(e.errors[0].errorCode).toBe('INVALID_CONTENT_TYPE');
      expect(e.errors[1].errorCode).toBe('NOT_SQUARE');
      expect(
        e.errors.map(validationError => {
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }

    expect(didError).toBe(true);
  });

  it('errors when file extension and content do not match up', async () => {
    let didError = false;
    try {
      await S.validateAssetsAsync({
        icon: './files/secretlyPng.jpg',
      });
    } catch (e) {
      didError = true;
      expect(e.errors[0].errorCode).toBe('FILE_EXTENSION_MISMATCH');
      expect(
        e.errors.map(validationError => {
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }

    expect(didError).toBe(true);
  });
});

describe('Holistic Unit Test', () => {
  it('good example app.json all', async () => {
    expect(await S.validateAll(good)).toEqual(undefined);
  });

  it('good example app.json schema', async () => {
    expect(await S.validateSchemaAsync(good)).toEqual(undefined);
  });

  it('bad example app.json schema', async () => {
    let didError = false;
    try {
      await S.validateSchemaAsync(bad);
    } catch (e) {
      didError = true;
      expect(e).toBeInstanceOf(SchemerError);
      const errors = e.errors;
      expect(errors.length).toBe(4);
      expect(
        errors.map(validationError => {
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });

  it('bad example app.json schema with field with not', async () => {
    let didError = false;
    try {
      await S.validateSchemaAsync(badWithNot);
    } catch (e) {
      didError = true;
      expect(e).toBeInstanceOf(SchemerError);
      const errors = e.errors;
      expect(errors.length).toBe(1);
      expect(
        errors.map(validationError => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });

  it('bad example app.json - invalid path for app icon', async () => {
    let didError = false;
    try {
      await S.validateAll(invalidAppIcon);
    } catch (e) {
      didError = true;
      expect(e).toBeInstanceOf(SchemerError);
      const errors = e.errors;
      expect(errors.length).toBe(1);
      expect(
        errors.map(validationError => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });
});

describe('Manual Validation Individual Unit Tests', () => {
  it('Local Icon', async () => {
    expect(await S.validateIcon('./files/check.png')).toEqual(undefined);
  });

  it('Local Square Icon correct', async () => {
    const S = new Schemer(
      { properties: { icon: { meta: { asset: true, square: true } } } },
      { rootDir: './__tests__' }
    );
    expect(await S.validateIcon('./files/check.png')).toEqual(undefined);
  });

  it('Local icon dimensions wrong', async () => {
    let didError = false;
    const S = new Schemer(
      {
        properties: {
          icon: {
            meta: {
              asset: true,
              dimensions: { width: 400, height: 401 },
              contentTypePattern: '^image/png$',
            },
          },
        },
      },
      { rootDir: './__tests__' }
    );
    try {
      await S.validateIcon('./files/check.png');
    } catch (e) {
      didError = true;
      expect(e).toBeTruthy();
      expect(e.errors.length).toBe(1);
      expect(
        e.errors.map(validationError => {
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });
});

describe('Individual Unit Tests', () => {
  it('Error when missing Required Property', async () => {
    let didError = false;
    const S = new Schemer(
      {
        properties: {
          name: {},
        },
        required: ['name'],
      },
      { rootDir: './__tests__' }
    );
    try {
      await S.validateAll({ noName: '' });
    } catch (e) {
      didError = true;
      expect(e.errors.length).toBe(1);
      expect(e.errors[0].errorCode).toBe(ErrorCodes.SCHEMA_MISSING_REQUIRED_PROPERTY);
      expect(
        e.errors.map(validationError => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });

  it('Error when data has an additional property', async () => {
    let didError = false;
    const S = new Schemer({ additionalProperties: false }, { rootDir: './__tests__' });
    try {
      await S.validateAll({ extraProperty: 'extra' });
    } catch (e) {
      didError = true;
      expect(e.errors.length).toBe(1);
      expect(e.errors[0].errorCode).toBe(ErrorCodes.SCHEMA_ADDITIONAL_PROPERTY);
      expect(
        e.errors.map(validationError => {
          // eslint-disable-next-line @typescript-eslint/no-unused-vars
          const { stack, ...rest } = validationError;
          return rest;
        })
      ).toMatchSnapshot();
    }
    expect(didError).toBe(true);
  });

  it.each`
    name            | expectedError
    ${'wilson'}     | ${undefined}
    ${[1, 2, 3, 4]} | ${'must be string'}
    ${23.232332}    | ${'must be string'}
    ${/regex.*/}    | ${'must be string'}
  `('validates name: $name', async ({ name, expectedError }) => {
    let didError = false;
    try {
      expect(await S.validateName(name)).toBe(undefined);
    } catch (e) {
      didError = true;
      expect(e.message).toBe(expectedError);
    }
    expect(didError).toBe(Boolean(expectedError));
  });

  it.each`
    slug                                             | expectedError
    ${'wilson'}                                      | ${undefined}
    ${12312123123}                                   | ${'must be string'}
    ${[1, 23]}                                       | ${'must be string'}
    ${'wilson123'}                                   | ${undefined}
    ${'wilson-123'}                                  | ${undefined}
    ${'wilson/test'}                                 | ${'\'\' must match pattern "^[a-zA-Z0-9_\\-]+$"'}
    ${'wilson-test%'}                                | ${'\'\' must match pattern "^[a-zA-Z0-9_\\-]+$"'}
    ${'wilson-test-zhao--javascript-is-super-funky'} | ${undefined}
  `('validates slug: $slug', async ({ slug, expectedError }) => {
    let didError = false;
    try {
      expect(await S.validateSlug(slug)).toBe(undefined);
    } catch (e) {
      didError = true;
      expect(e.message).toBe(expectedError);
    }
    expect(didError).toBe(Boolean(expectedError));
  });

  it.each`
    sdkVersion       | expectedError
    ${'1.0.0'}       | ${undefined}
    ${'2.0.0.0.1'}   | ${undefined}
    ${'UNVERSIONED'} | ${undefined}
    ${'12.2a.3'}     | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
    ${'9,9,9'}       | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
    ${'1.2'}         | ${'\'\' must match pattern "^(\\d+\\.\\d+\\.\\d+)|(UNVERSIONED)$"'}
  `('validates SDK version: $sdkVersion', async ({ sdkVersion, expectedError }) => {
    let didError = false;
    try {
      expect(await S.validateSdkVersion(sdkVersion)).toBe(undefined);
    } catch (e) {
      didError = true;
      expect(e.message).toBe(expectedError);
    }
    expect(didError).toBe(Boolean(expectedError));
  });
});
