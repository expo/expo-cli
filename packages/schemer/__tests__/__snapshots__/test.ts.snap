// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Holistic Unit Test bad example app.json - invalid path for app icon 1`] = `
Array [
  Object {
    "data": "./unknown/path.png",
    "errorCode": "INVALID_ASSET_URI",
    "fieldPath": "icon",
    "message": "cannot access file at './unknown/path.png'",
    "meta": Object {
      "asset": true,
      "bareWorkflow": "To change your app's icon, edit or replace the files in \`ios/<PROJECT-NAME>/Assets.xcassets/AppIcon.appiconset\` (we recommend using Xcode), and \`android/app/src/main/res/mipmap-<RESOLUTION>\`. Be sure to follow the guidelines for each platform ([iOS](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/app-icon/), [Android 7.1 and below](https://material.io/design/iconography/#icon-treatments), and [Android 8+](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive)) and to provide your new icon in each existing size.",
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Holistic Unit Test bad example app.json schema 1`] = `
Array [
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_MISSING_REQUIRED_PROPERTY",
    "fieldPath": "",
    "message": "is missing required property 'name'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'orientaasdfasdftion'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": Object {
      "asdfasdfandroid": Object {
        "package": "com.yourcompany.yourappname",
      },
      "icon": "DoesNotExist.png",
      "orientaasdfasdftion": "portrait",
      "sdkVersion": "17.0.0abad",
      "slug": "1*@)#($*@)(#$*)",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'asdfasdfandroid'",
    "meta": undefined,
    "name": "ValidationError",
  },
  Object {
    "data": "1*@)#($*@)(#$*)",
    "errorCode": "SCHEMA_INVALID_PATTERN",
    "fieldPath": "slug",
    "message": "'slug' must match pattern \\"^[a-zA-Z0-9_\\\\-]+$\\"",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Holistic Unit Test bad example app.json schema with field with not 1`] = `
Array [
  Object {
    "data": "1.0",
    "errorCode": "SCHEMA_INVALID_NOT",
    "fieldPath": "runtimeVersion",
    "message": "'runtimeVersion' should be not a decimal ending in a 0.",
    "meta": Object {
      "notHuman": "Not a decimal ending in a 0.",
    },
    "name": "ValidationError",
  },
]
`;

exports[`Image Validation errors for webp images 1`] = `
Array [
  Object {
    "data": "./files/webp.webp",
    "errorCode": "INVALID_CONTENT_TYPE",
    "fieldPath": "Android.adaptiveIcon.foregroundImage",
    "message": "field 'Android.adaptiveIcon.foregroundImage' should point to .png image but the file at './files/webp.webp' has type webp",
    "meta": Object {
      "asset": true,
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
  Object {
    "data": "./files/webp.webp",
    "errorCode": "NOT_SQUARE",
    "fieldPath": "Android.adaptiveIcon.foregroundImage",
    "message": "image should be square, but the file at './files/webp.webp' has dimensions 320x214",
    "meta": Object {
      "asset": true,
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Image Validation errors when file extension and content do not match up 1`] = `
Array [
  Object {
    "data": "./files/secretlyPng.jpg",
    "errorCode": "FILE_EXTENSION_MISMATCH",
    "fieldPath": "icon",
    "message": "the file extension should match the content, but the file extension is .jpg while the file content at './files/secretlyPng.jpg' is of type png",
    "meta": Object {
      "asset": true,
      "bareWorkflow": "To change your app's icon, edit or replace the files in \`ios/<PROJECT-NAME>/Assets.xcassets/AppIcon.appiconset\` (we recommend using Xcode), and \`android/app/src/main/res/mipmap-<RESOLUTION>\`. Be sure to follow the guidelines for each platform ([iOS](https://developer.apple.com/design/human-interface-guidelines/ios/icons-and-images/app-icon/), [Android 7.1 and below](https://material.io/design/iconography/#icon-treatments), and [Android 8+](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive)) and to provide your new icon in each existing size.",
      "contentTypeHuman": ".png image",
      "contentTypePattern": "^image/png$",
      "square": true,
    },
    "name": "ValidationError",
  },
]
`;

exports[`Individual Unit Tests Error when data has an additional property 1`] = `
Array [
  Object {
    "data": Object {
      "extraProperty": "extra",
    },
    "errorCode": "SCHEMA_ADDITIONAL_PROPERTY",
    "fieldPath": "",
    "message": "should NOT have additional property 'extraProperty'",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Individual Unit Tests Error when missing Required Property 1`] = `
Array [
  Object {
    "data": Object {
      "noName": "",
    },
    "errorCode": "SCHEMA_MISSING_REQUIRED_PROPERTY",
    "fieldPath": "",
    "message": "is missing required property 'name'",
    "meta": undefined,
    "name": "ValidationError",
  },
]
`;

exports[`Manual Validation Individual Unit Tests Local icon dimensions wrong 1`] = `
Array [
  Object {
    "data": "./files/check.png",
    "errorCode": "INVALID_DIMENSIONS",
    "fieldPath": "icon",
    "message": "'icon' should have dimensions 400x401, but the file at './files/check.png' has dimensions 512x512",
    "meta": Object {
      "asset": true,
      "contentTypePattern": "^image/png$",
      "dimensions": Object {
        "height": 401,
        "width": 400,
      },
    },
    "name": "ValidationError",
  },
]
`;
